update help output

vlada-shubina opened the pull request:
fixes dotnet/templating#2591 update help output

updated usage
updated options with arguments
updated desciriptions (mainly from docs.microsoft.com)

due to CommandLine library doesn't support the order of arguments used in template engine, had to do own implementation

vlada-shubina commented on src/Microsoft.TemplateEngine.Cli/LocalizableStrings.resx:
```diff
@@ -682,4 +691,27 @@ Run 'dotnet {1} --show-aliases' with no args to show all aliases.</value>
+    <value>Arguments:</value>
+  </data>
+  <data name="CommandUsage" xml:space="preserve">
+    <value>Usage: 
```
@KathleenDollard 
I put the usage as below. Note: it is not similar to other dotnet tools, however it is more descriptive for user. 

https://docs.microsoft.com/dotnet/core/tools/dotnet-new uses similar way.

donJoseLuis commented on src/Microsoft.TemplateEngine.Cli/CommandParsing/CommandLineHelpViewer.cs:
```diff
@@ -0,0 +1,120 @@
+{
+    internal static class CommandLineHelpViewer
+    {
+        private static int columnGutterWidth = 3;
```
consider PASCAL naming for constants\static fields.
https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-1.1/d53b55ey(v=vs.71)

donJoseLuis commented on src/Microsoft.TemplateEngine.Cli/CommandParsing/CommandLineHelpViewer.cs:
```diff
@@ -0,0 +1,120 @@
+    {
+        private static int columnGutterWidth = 3;
+        private const string templateArgument = "TEMPLATE";
+        internal static string HelpView(this Command newCommand)
```
please separate fields and methods with a line

donJoseLuis commented on src/Microsoft.TemplateEngine.Cli/CommandParsing/CommandLineHelpViewer.cs:
```diff
@@ -0,0 +1,120 @@
+        }
+
+        private static void WriteOptionsSection(
+            Command command,
```
minor: consider homogeneous parameter order.  In some peer methods, for the same parameters order is
- StringBuilder, Command
- here it is Command, StringBuilder

donJoseLuis commented on src/Microsoft.TemplateEngine.Cli/CommandParsing/CommandLineHelpViewer.cs:
```diff
@@ -0,0 +1,120 @@
+
+        private static string LeftColumnText(Option option)
+        {
+            var leftColumnText = "  " +
```
minor: please homogenize style.  This indentation stands out, unsure why.

donJoseLuis commented on src/Microsoft.TemplateEngine.Cli/CommandParsing/CommandLineHelpViewer.cs:
```diff
@@ -0,0 +1,120 @@
+
+namespace Microsoft.TemplateEngine.Cli.CommandParsing
+{
+    internal static class CommandLineHelpViewer
```
minor: if this is an extension class, consider the suffix "Extension" for the class name

AR-May commented on src/Microsoft.TemplateEngine.Cli/LocalizableStrings.resx:
```diff
@@ -187,16 +187,17 @@
   <data name="DestructiveChangesNotification" xml:space="preserve">
     <value>Creating this template will make changes to existing files:</value>
   </data>
-  <data name="DisplaysHelp" xml:space="preserve">
```
We previously agreed not to remove yet the localized strings that are already there. 
Should we reconsider the decision with our new knowledge about localization?

vlada-shubina replied:
I can keep old names, however the names are very different for each option:
- DisplaysHelp is used for --help option description
- ShowFilteredTemplates is used for --type option description 
and so on.
It's much convenient to have them unified.

vlada-shubina commented:
@KathleenDollard: updated PR with yesterday suggestions:
- made usage similar to other dotnet commands
- added previous version of usage as "Examples" section
- fixed description for --nuget-source option.

Please review

KathleenDollard commented:
I really like what you did here. I am a little worried about the length, so I have some suggestions to shorten a few things up. Also, does this wrap, and how does it look wrapped?

Allows the internal dotnet restore command to stop and wait for user input or action (for example to complete authentication).
Allows the implicit restore to stop and wait for user input if needed (for example to complete authentication).
_Can the user perform any action other than input?_

-lang, --language <LANGUAGE>      Specifies the template language to use.
                                    When using with dotnet new --list or -l, filters templates based on language.
-lang, --language <LANGUAGE>      Specifies the template language to use for template selection or filtering.

--type <TYPE>                     Specifies the template type to use  for template selection or filtering. Predefined values are "project", "item" or "other".

 -u, --uninstall <PATH|NUGET_ID>   Uninstalls a template pack at the path or using NuGet ID provided.
                                    When the argument value is not specified, currently installed template packs and their templates are displayed.
                                    If the package was installed by either a NuGet feed or by a .nupkg file directly, provide the identifier without  the version number.
                                    If the package was installed by specifying a path to the .template.config folder, use that absolute path to uninstall the package without a terminating slash. You can see this path with the dotnet new -u command. 

 I assume this will just be new in the final: _dotnet new3_

vlada-shubina commented:
Thanks @KathleenDollard 

_--interactive option, can the user perform any action other than input?_ 
to be honest I'm not sure, it is exactly the same as dotnet restore --interactive https://docs.microsoft.com/ru-ru/dotnet/core/tools/dotnet-restore - dotnet restore is run during template installation.  As far as I know the common use of it is to enter credentials required for restore from private nuget feed.

new3 will become new after insertion to sdk

Line wrapping:
![image](https://user-images.githubusercontent.com/56045248/98115072-e597b680-1eae-11eb-862c-a7f8ce34c094.png)
The text is wrapped for --install and --uninstall, it's not nice but readable (same for other dotnet commands). As install and uninstall is not trivial my preference would be not to shorten it for nicer UX. 

Updated output is below (note it has added options and minor changes now after rebase)
```
> dotnet .\dotnet-new3.dll --help
Usage: dotnet new3 <TEMPLATE> [options]

Arguments:
  <TEMPLATE>        The template to instantiate when the command is invoked. Each template might have specific options you can pass.
                    When using with dotnet new --list or -l, <TEMPLATE> is used to filter the templates which match on the template name.
                    When using with dotnet new --help or -h, the information about <TEMPLATE> and its template specific options is displayed.
                    Should not be used with -i, --install, -u, --uninstall, --update-check, --update-apply.

Options:
  -n, --name <OUTPUT_NAME>          The name for the output being created. If no name is specified, the name of the output directory is used.
  -o, --output <OUTPUT_DIRECTORY>   Location to place the generated output. The default is the current directory.
  --interactive                     Allows the implicit restore to stop and wait for user input if needed (for example to complete authentication)
  --dry-run                         Displays a summary of what would happen if the given command line were run if it would result in a template creation.
  --force                           Forces content to be generated even if it would change existing files.
  -lang, --language <LANGUAGE>      Specifies the template language to use for template selection or filtering.
  --type <TYPE>                     Specifies the template type to use for template selection or filtering. Predefined values are "project" and "item".
  -h, --help                        Shows help information.
                                    If template name is specified, shows information about template and template specific options.
  -l, --list                        Lists templates containing the specified template name. If no name is specified, lists all templates.
  -i, --install <PATH|NUGET_ID>     Installs a template pack from the path or using NuGet ID provided.
                                    To install the package from the NuGet feed, use the NuGet package identfier. Use "package-identifier::version" to specify the version to install. If a version of the template was already installed when you run this command, the template will be updated to the specified version, or to the latest stable version if no version was specified.
                                    To install the package from the file system, specify the path to the NuGet package file.
                                    To install the template from the file system directory, specify the folder path of the .template.config folder. Do not include a final terminating directory slash on your template path.
  --nuget-source <SOURCE>           Specifies a NuGet source to use during install. Applicable only with -i or --install option.
  -u, --uninstall <PATH|NUGET_ID>   Uninstalls a template pack at the path or using NuGet ID provided.
                                    When the argument value is not specified, currently installed template packs and their templates are displayed.
                                    If the package was installed by either a NuGet feed or by a .nupkg file directly, provide the identifier without the version number.
                                    If the package was installed by specifying a path to the .template.config folder, use that absolute path to uninstall the package without a terminating slash. You can see this path with the dotnet new -u command.
  --update-check                    Checks the currently installed template packs for updates.
  --update-apply                    Checks the currently installed template packs for update, and install the updates.
  --columns <COLUMNS_LIST>          Comma separated list of columns to show for --list or -l option.  The supported columns are:
                                    - language - comma separated list of languages supported by the template
                                    - tags - the list of template tags
                                    - author - the template author
                                    - type - the template type: project or item
                                    The template name and short name are shown always.
                                    The default list of columns shown without the option: template name, short name, language, tags; equivalent to --columns=language,tags.
  --columns-all                     Shows all available columns for --list or -l option, equivalent to --columns=language,tags,author,type.

Examples:
        dotnet new3 <TEMPLATE> [-n|--name <OUTPUT_NAME>] [-o|--output <OUTPUT_DIRECTORY>] [--interactive] [--dry-run] [--force] [--language|-lang <LANGUAGE>] [--type|-t <TYPE>] [template options]
        dotnet new3 --help|-h
        dotnet new3 <TEMPLATE>  --help|-h [--language|-lang <LANGUAGE>] [--type|-t <TYPE>]
        dotnet new3 [TEMPLATE]  --list|-l [--language|-lang <LANGUAGE>] [--type|-t <TYPE>] [--columns-all] [--columns <COLUMNS_LIST>]
        dotnet new3 -i|--install <PATH|NUGET_ID> [--nuget-source <SOURCE>]
        dotnet new3 -u|--uninstall [PATH|NUGET_ID]
        dotnet new3 --update-check
        dotnet new3 --update-apply

```

donJoseLuis commented on src/Microsoft.TemplateEngine.Cli/CommandParsing/CommandLineHelpViewerExtensions.cs:
```diff
@@ -0,0 +1,111 @@
+
+        private static void WriteSynopsys(StringBuilder helpView, Command command)
+        {
+            helpView.AppendLine(string.Format(LocalizableStrings.NewCommandUsage, command.Name, TemplateArgument));
```
minor:  StringBuilder Append, return a corresponding Stringbuilder instance.   Since you're not consuming this, consider...
`_ = helpView.AppendLine(string.Format(LocalizableStrings.NewCommandUsage, command.Name, TemplateArgument));`

bozturkMSFT commented on src/Microsoft.TemplateEngine.Cli/CommandParsing/CommandLineHelpViewerExtensions.cs:
```diff
@@ -0,0 +1,111 @@
+            Option[] options = command
+                .DefinedOptions
+                .Where(o => !o.IsHidden())
+                .ToArray();
```
If an array is not a must, `ToList` maybe a bit more efficient here, because the allocated memory doesn't need to exactly match the size of the data.

vlada-shubina replied:
Most of the members of this class are copied from [Microsoft.DotNet.Cli.CommandLine](https://github.com/dotnet/CliCommandLineParser/blob/master/src/source/CommandLine/HelpViewExtensions.cs), including the one in this and other comments. Including this code was the cost of getting proper output for dotnet new --help. I had to change implementation as Microsoft.DotNet.Cli.CommandLine do not support arguments before options and there is no way we can get proper usage with it.  However, options output is correct and should stay as is. Due to the fact that these methods (WriteOptionsSection) are private it was impossible to reuse them when changing the HelpView() to add missing Arguments description, so I had to add them to Template Engine. 
Since we are considering to move to System.CommandLine ASAP when we are back to work on Template Engine, probably this extension will go away very soon. Therefore, I'm not sure if we should invest in improving this code. 
I would like to reconsider the decision whether to refactor this class, once we get approval to include the change to 5.0.200. If won't have the approval, high chance that move to System.CommandLine will happen sooner and this code will be dropped.

bozturkMSFT commented on src/Microsoft.TemplateEngine.Cli/CommandParsing/CommandLineHelpViewerExtensions.cs:
```diff
@@ -0,0 +1,111 @@
+            int leftColumnWidth = leftColumnTextFor
+                                      .Values
+                                      .Select(s => s.Length)
+                                      .OrderBy(length => length)
```
We can just use `Max` extension method from `Linq` to get the string with the largest length. Sorting can be avoided in that case.
Or am I just misunderstanding the intent here?

bozturkMSFT commented on src/Microsoft.TemplateEngine.Cli/LocalizableStrings.resx:
```diff
@@ -215,8 +216,11 @@
+  <data name="OptionDescriptionInstall" xml:space="preserve">
+    <value>Installs a template pack from the path or using NuGet ID provided.
+To install the package from the NuGet feed, use the NuGet package identfier. Use "package-identifier::version" to specify the version to install. If a version of the template was already installed when you run this command, the template will be updated to the specified version, or to the latest stable version if no version was specified.
+To install the package from the file system, specify the path to the NuGet package file.
```
Would it help to mention the expected file extension here?

vlada-shubina replied:
I think it would be helpful, added.

vlada-shubina commented:
Closing for now, might come back to it later if needed. Should be improved when new CLI parser is implemented anyway.
